### PR TITLE
dek_normalizeFasta_IT

### PR DESCRIPTION
**Pull request by dekling**
Completed integration test and added two test files.

### CODE REVIEW

**Comment by dekling** [comment]
Completed integration test and added a couple of test files.

File: src/test/java/org/broadinstitute/hellbender/tools/picard/fasta/dek_Normalizefasta_IT.java [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+import org.broadinstitute.hellbender.utils.test.BaseTest;
+import org.testng.annotations.Test;
+import java.io.File;
+import java.io.IOException;
+
+/**
+ * Created by dkling on 9/2/15.
+ */
+
+    public final class dek_Normalizefasta_IT extends CommandLineProgramTest {
```
Comment (akiezun) [inline src/test/java/org/broadinstitute/hellbender/tools/picard/fasta/dek_Normalizefasta_IT.java]: rename this class to NormalizeFastaIntegrationTest`

File: src/test/java/org/broadinstitute/hellbender/utils/fasta/CachingIndexedFastaSequenceFileUnitTest.java [hunk 2]
```diff
@@ -1,9 +1,7 @@
 package org.broadinstitute.hellbender.utils.fasta;
```
Comment (akiezun) [inline src/test/java/org/broadinstitute/hellbender/utils/fasta/CachingIndexedFastaSequenceFileUnitTest.java]: Is this file actually changed in this branch? If not, remove the changes from the branch. if yes, explain the changes.

File: src/test/resources/org/broadinstitute/hellbender/tools/picard/fasta/Normalizedexp.fasta [hunk 3]
```diff
@@ -0,0 +1,1001 @@
+>chr1
```
Comment (akiezun) [inline src/test/resources/org/broadinstitute/hellbender/tools/picard/fasta/Normalizedexp.fasta]: use a much smaller input file. 10 lines will suffice

File: src/test/java/org/broadinstitute/hellbender/tools/picard/fasta/dek_Normalizefasta_IT.java [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+        @Test
+        public void testNormalize () throws IOException {
+
+            final File input = new File(TEST_DATA_PATH, "exampleFASTA.fasta");
+            final File expectedFile = new File(TEST_DATA_PATH, "Normalizedexp.fasta");
+            final File outfile = BaseTest.createTempFile("normalized", ".fasta");
+
+            final String[] args = {
+                    "--INPUT", input.getAbsolutePath(),
+                    "--OUTPUT", outfile.getAbsolutePath(),
```
Comment (akiezun) [inline src/test/java/org/broadinstitute/hellbender/tools/picard/fasta/dek_Normalizefasta_IT.java]: use the LINE_LENGTH option, set to a small number like 5

File: src/test/java/org/broadinstitute/hellbender/tools/picard/fasta/dek_Normalizefasta_IT.java [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+            };
+
+            runCommandLine(args);
+            IntegrationTestSpec.assertEqualTextFiles(outfile, expectedFile, "#");
+
+        }
+
```
Comment (akiezun) [inline src/test/java/org/broadinstitute/hellbender/tools/picard/fasta/dek_Normalizefasta_IT.java]: add a test that uses LINE_LENGTH set to null

Comment (akiezun) [inline src/test/java/org/broadinstitute/hellbender/tools/picard/fasta/dek_Normalizefasta_IT.java]: add another test that uses `TRUNCATE_SEQUENCE_NAMES_AT_WHITESPACE=true`. Use an input file that has spaces in sequence names, eg `Chromosome 1`

**Comment by akiezun** [comment]
looks fine, small things to change - use smaller input file and exercise more options of the tool. Back to @dekling. Please reassign to me when you're done

File: src/test/resources/org/broadinstitute/hellbender/tools/picard/fasta/Normalizedexp.fasta (hunk 3 above)
Comment (dekling) [inline src/test/resources/org/broadinstitute/hellbender/tools/picard/fasta/Normalizedexp.fasta]: done

File: src/test/java/org/broadinstitute/hellbender/tools/picard/fasta/dek_Normalizefasta_IT.java (hunk 4 above)
Comment (dekling) [inline src/test/java/org/broadinstitute/hellbender/tools/picard/fasta/dek_Normalizefasta_IT.java]: done

File: src/test/java/org/broadinstitute/hellbender/tools/picard/fasta/dek_Normalizefasta_IT.java (hunk 5 above)
Comment (dekling) [inline src/test/java/org/broadinstitute/hellbender/tools/picard/fasta/dek_Normalizefasta_IT.java]: I am a bit confused.  When I add the argument "LINE_LENGTH=null" in Picard v. 1.138, I get the error, "Can not set int field picard.reference.NormalizeFasta.LINE_LENGTH to null value".  Should we expect the same in Hellbender?

File: src/test/java/org/broadinstitute/hellbender/utils/fasta/CachingIndexedFastaSequenceFileUnitTest.java (hunk 2 above)
Comment (dekling) [inline src/test/java/org/broadinstitute/hellbender/utils/fasta/CachingIndexedFastaSequenceFileUnitTest.java]: I had accidentally changed/moved it but subsequently placed it back in its place.

File: src/test/java/org/broadinstitute/hellbender/tools/picard/fasta/dek_Normalizefasta_IT.java (hunk 5 above)
Comment (dekling) [inline src/test/java/org/broadinstitute/hellbender/tools/picard/fasta/dek_Normalizefasta_IT.java]: Please clarify what you mean by "not passing...".  I added (possibly back in) "(expectedExceptions = UserException.BadArgumentValue.class)" after @Test.  This enabled Hellbender to handle LIne_Length=null.  If I remove the code containing Line_Length=null, I also have to remove, " (expectedExceptions = UserException.BadArgumentValue.class)" after @Test, in order for the program to run.

Comment (dekling) [inline src/test/java/org/broadinstitute/hellbender/tools/picard/fasta/dek_Normalizefasta_IT.java]: Added another test that uses TRUNCATE_SEQUENCE_NAMES_AT_WHITESPACE=true. Used an input file that has spaces in sequence names, eg Chromosome 1.  Worked OK but had to remove (expectedExceptions = UserException.BadArgumentValue.class)" after @Test, in order for the program to run.  Gives green light but still outputs a verbosity error.

Comment (dekling) [inline src/test/java/org/broadinstitute/hellbender/tools/picard/fasta/dek_Normalizefasta_IT.java]: Got everything working.  Then retested LINE_LENGTH=NULL.  Threw the following error:


java.lang.IllegalArgumentException: Can not set int field org.broadinstitute.hellbender.tools.picard.fastq.NormalizeFasta.LINE_LENGTH to null value
	at sun.reflect.UnsafeFieldAccessorImpl.throwSetIllegalArgumentException(UnsafeFieldAccessorImpl.java:167)
	at sun.reflect.UnsafeFieldAccessorImpl.throwSetIllegalArgumentException(UnsafeFieldAccessorImpl.java:171)
	at sun.reflect.UnsafeIntegerFieldAccessorImpl.set(UnsafeIntegerFieldAccessorImpl.java:80)
	at java.lang.reflect.Field.set(Field.java:764)
	at org.broadinstitute.hellbender.cmdline.CommandLineParser$ArgumentDefinition.setFieldValue(CommandLineParser.java:838)
	at org.broadinstitute.hellbender.cmdline.CommandLineParser.setArgument(CommandLineParser.java:402)
	at org.broadinstitute.hellbender.cmdline.CommandLineParser.parseArguments(CommandLineParser.java:268)
	at org.broadinstitute.hellbender.cmdline.CommandLineProgram.parseArgs(CommandLineProgram.java:183)
	at org.broadinstitute.hellbender.cmdline.CommandLineProgram.instanceMain(CommandLineProgram.java:106)
	at org.broadinstitute.hellbender.Main.instanceMain(Main.java:71)
	at org.broadinstitute.hellbender.Main.instanceMain(Main.java:78)
	at org.broadinstitute.hellbender.CommandLineProgramTest.runCommandLine(CommandLineProgramTest.java:75)
	at org.broadinstitute.hellbender.tools.picard.fastq.NormalizeFastaIntegrationTest.testNormalize(NormalizeFastaIntegrationTest.java:38)

Process finished with exit code 0

File: src/test/java/org/broadinstitute/hellbender/tools/picard/analysis/QualityScoreDistributionIntegrationTest.java [hunk 6]
```diff
@@ -14,7 +14,7 @@
 public class QualityScoreDistributionIntegrationTest extends CommandLineProgramTest {
     private static final File TEST_DATA_DIR = new File(getTestDataDir(), "picard/analysis/QualityScoreDistribution");
 
-    @Test(groups = {"R"})
+    @Test
```
Comment (akiezun) [inline src/test/java/org/broadinstitute/hellbender/tools/picard/analysis/QualityScoreDistributionIntegrationTest.java]: why was the R group removed? please restore

File: src/test/java/org/broadinstitute/hellbender/utils/fasta/CachingIndexedFastaSequenceFileUnitTest.java (hunk 2 above)
Comment (akiezun) [inline src/test/java/org/broadinstitute/hellbender/utils/fasta/CachingIndexedFastaSequenceFileUnitTest.java]: please restore the original version

File: src/test/resources/org/broadinstitute/hellbender/tools/picard/fasta/Normalizedexp.fasta (hunk 3 above)
Comment (akiezun) [inline src/test/resources/org/broadinstitute/hellbender/tools/picard/fasta/Normalizedexp.fasta]: remove this file then

File: src/test/resources/org/broadinstitute/hellbender/tools/picard/fasta/norm_trunc_fasta.fasta [hunk 7]
```diff
@@ -0,0 +1,8 @@
+>chr1
```
Comment (akiezun) [inline src/test/resources/org/broadinstitute/hellbender/tools/picard/fasta/norm_trunc_fasta.fasta]: is this the actual input used? if yes, it should go to a directory picard/fasta/fastq/NormalizeFasta/

if not, remove this file

File: src/test/resources/org/broadinstitute/hellbender/tools/picard/fasta/testFASTA_WS_4.fasta [hunk 8]
```diff
@@ -0,0 +1,284 @@
+>Chromosome
```
Comment (akiezun) [inline src/test/resources/org/broadinstitute/hellbender/tools/picard/fasta/testFASTA_WS_4.fasta]: is this the actual input used? if yes, it should go to a directory picard/fasta/fastq/NormalizeFasta/

if not, remove this file

File: src/test/resources/org/broadinstitute/hellbender/tools/picard/fasta/testfastq.fastq [hunk 9]
```diff
@@ -0,0 +1,55 @@
+@20FUKAAXX100202:6:27:4968:125377
```
Comment (akiezun) [inline src/test/resources/org/broadinstitute/hellbender/tools/picard/fasta/testfastq.fastq]: what is this file for? please remove it

**Comment by akiezun** [comment]
please clean up this pull request - there should be only 3 files:
testfasta.fasta 
testFASTA_WS_4.fasta
and NormalizeFastaIntegrationTest.java

(the first 2 should be moved to src/test/resources/org/broadinstitute/hellbender/tools/picard/fasta/NormalizeFasta/)

**Comment by akiezun** [comment]
back to @dekling

File: src/test/resources/org/broadinstitute/hellbender/tools/picard/fasta/Normalizedexp.fasta (hunk 3 above)
Comment (dekling) [inline src/test/resources/org/broadinstitute/hellbender/tools/picard/fasta/Normalizedexp.fasta]: done.

File: src/test/resources/org/broadinstitute/hellbender/tools/picard/fasta/testFASTA_WS_4.fasta (hunk 8 above)
Comment (dekling) [inline src/test/resources/org/broadinstitute/hellbender/tools/picard/fasta/testFASTA_WS_4.fasta]: This is the output file.

File: src/test/java/org/broadinstitute/hellbender/tools/picard/fastq/NormalizeFastaIntegrationTest.java [hunk 10]
```diff
@@ -0,0 +1,50 @@
...
+import org.testng.annotations.Test;
+
+import java.io.File;
+import java.io.IOException;
+
+/**
+ * Created by dkling on 9/3/15.
+ */
+
+    public final class NormalizeFastaIntegrationTest extends CommandLineProgramTest {
```
Comment (akiezun) [inline src/test/java/org/broadinstitute/hellbender/tools/picard/fastq/NormalizeFastaIntegrationTest.java]: remove unnecessary empty spaces and empty lines

File: src/test/java/org/broadinstitute/hellbender/tools/picard/fastq/NormalizeFastaIntegrationTest.java [hunk 11]
```diff
@@ -0,0 +1,50 @@
...
+                    "--TRUNCATE_SEQUENCE_NAMES_AT_WHITESPACE", "TRUE",
+                    "--LINE_LENGTH", "5",
+            };
+
+            runCommandLine(args);
+
+
+
+
+            IntegrationTestSpec.assertEqualTextFiles(outfile, expectedFile, "#");
```
Comment (akiezun) [inline src/test/java/org/broadinstitute/hellbender/tools/picard/fastq/NormalizeFastaIntegrationTest.java]: change to `IntegrationTestSpec.assertEqualTextFiles(outfile, expectedFile);`

File: src/test/resources/org/broadinstitute/hellbender/tools/picard/fasta/testFASTA_WS_4.fasta (hunk 8 above)
Comment (akiezun) [inline src/test/resources/org/broadinstitute/hellbender/tools/picard/fasta/testFASTA_WS_4.fasta]: create a directory `src/test/resources/org/broadinstitute/hellbender/tools/picard/fasta/NormalizeFastaIntegration`
and move this file there. And change the code in NormalizeFastaIntegrationTest.java accordingly (ie change "picard/fasta/"  to "picard/fasta/NormalizeFastaIntegration" on line 17)

File: src/test/resources/org/broadinstitute/hellbender/tools/picard/fasta/testfasta.fasta [hunk 12]
```diff
@@ -0,0 +1,15 @@
+>Chromosome 20
```
Comment (akiezun) [inline src/test/resources/org/broadinstitute/hellbender/tools/picard/fasta/testfasta.fasta]: create a directory src/test/resources/org/broadinstitute/hellbender/tools/picard/fasta/NormalizeFastaIntegration
and move this file there.

**Comment by akiezun** [comment]
excellent. very close. Just a few minor edits and it's done. back to @dekling
